rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,364 @@
+- Decouples input from focus management
+- Decouples focus management from `bevy_ui`
+- User-specified input handling for UI interaction
+- An optional menu tree management system
+
+## Terminology
+
+- _menu tree_: the access hierarchy of menus.
+- _activating_ a focusable: Sending the `Action` event while this focusable is focused.
+- _focused menu_: The menu in which the currently focused `Focusable` is.
+- _child menu_ (of a focusable): The menu accessed from this focusable.
+
+## Motivation
+
+`bevy_ui` interaction story is crude.
+The only way to interact with the UI currently
+is through the `Interaction` and `FocusPolicy` components.
+`Interaction` only supports mouse or touch-based input.
+We want gamepad supports and more game-oriented functionalities,
+such as menu trees and movement between menus.
+
+The current UI focus system is limited to `bevy_ui` and
+can't be swapped out for something the user deems more appropriate.
+
+This is in contradiction with the general philosophy of bevy,
+which is of providing to the user all the tools necessary to create something
+that fits their needs optimally.
+
+This RFC proposes to resolve all those questions.
+[`bevy-ui-navigation`] provides a highly customizable ECS-based navigation engine.
+
+## User-facing explanation
+
+The [bevy-ui-navigation README][`bevy-ui-navigation`] is a good start.
+To summarize, here are the headlines:
+- The navigation system can only be interacted with with events:
+  `NavRequest` as input and `NavEvent` as output.
+- `NavEvent::FocusChanged` contains the list of entities that are traversed
+  to go from the last focus position to the new one.
+- ui-navigation provides default input systems, 
+  but it's possible to disable them and instead use custom ones.
+  This is why we restrict interactions with navigation to events.
+- ui-navigation provides a system to declare buttons as "leading to menu X"
+  in order to create menu trees.
+- ui-navigation defines [custom `SystemParam`][event_helpers]s
+  to ease reaction to events (such a button press or menu change).
+  The most naive usage results into [elm architecture][elm_architecture]-like code.
+  Systems updating the model (ECS hierarchy) and systems reacting to events. 
+- It is possible to create isolated menus by using the `NavMenu` component.
+- All `Focusable` children in the `NavMenu` entity's tree will be part
+  of this menu, and moving using a gamepad from one focusable in a menu
+  can only lead to another focusable within the same menu.
+  You can specify a focusable in any other menu that will lead to this
+  one once activated.
+
+## Implementation
+
+Since `bevy-ui-navigation` already exists,
+we will discuss its design, it's functionalities and their uses.
+
+Implementation details are already covered by the implementation,
+which btw is fully documented (including internal architecture).
+
+### Exposed API
+
+The crux of `bevy-ui-navigation` are the following types:
+- `Focusable` component
```

Comment:
Can we model this as an enum?

Reply:
Interesting ideas. The current design has `Focusable` be two different enums (one enum for the state `FocusState`, and one enum for the "kind" of focusable (lock, cancel, normal)). It might be possible to separate them in two components. I've never considered this.

However, we need some sort of strategy to prevent users from directly mutating the state, because the navigation algorithm relies on having full control over it. It's read only (and for very good reason!)